### PR TITLE
[WIP] Add AMP Conversion support for BucketingModule

### PR DESCRIPTION
**Pull request by anirudh2290**
## Description ##
- Add conversion APIs for bucketing module
- Add support for saving and loading BucketingModule from a checkpoint
- Add support for loading BucketingModule from a dict of bucket keys to symbols
- Add tests

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by karan6181** [comment]
@mxnet-label-bot add [Module, Python, pr-work-in-progress]

File: python/mxnet/model.py [hunk 1]
```diff
@@ -423,6 +423,20 @@ def save_checkpoint(prefix, epoch, symbol, arg_params, aux_params, remove_amp_ca
     logging.info('Saved checkpoint to \"%s\"', param_name)
 
 
+def load_params(prefix, epoch):
```
Comment (larroy) [inline python/mxnet/model.py]: nice

File: python/mxnet/module/bucketing_module.py [hunk 2]
```diff
@@ -335,12 +339,16 @@ def bind(self, data_shapes, label_shapes=None, for_training=True,
...
 
         symbol, data_names, label_names = self._call_sym_gen(self._default_bucket_key)
-        module = Module(symbol, data_names, label_names, logger=self.logger,
-                        context=self._context, work_load_list=self._work_load_list,
-                        fixed_param_names=self._fixed_param_names,
-                        state_names=self._state_names,
-                        group2ctxs=self._group2ctxs,
-                        compression_params=self._compression_params)
+        module = None
+        if not self._default_bucket_key in self._buckets:
```
Comment (larroy) [inline python/mxnet/module/bucketing_module.py]: do you have to check that _default_bucket_key is set? If it's none it shouldn't be in the dictionary I assume.

File: tests/python/train/test_bucketing.py [hunk 3]
```diff
@@ -20,9 +20,32 @@
...
+from mxnet.contrib.amp import amp
 
 
-def test_bucket_module():
+def prepare_bucketing_data(buckets, len_vocab, batch_size, invalid_label, num_sentence):
+    train_sent = []
+    val_sent = []
+
+    for _ in range(num_sentence):
+        len_sentence = randint(6, max(buckets)-1) # leave out the two last buckets empty
```
Comment (larroy) [inline tests/python/train/test_bucketing.py]: Is this comment correct? Randint doc says:
```
randint(a, b) method of random.Random instance
    Return random integer in range [a, b], including both end points.
```
So b is included in the range. Or am I missing something?

File: example/rnn/bucketing/cudnn_rnn_bucketing.py [hunk 4]
```diff
@@ -67,6 +68,10 @@
                     help='dropout probability (1.0 - keep probability)')
 parser.add_argument('--rnntype', type=str, default='lstm',
                     help='rnn type: gru, lstm, rnn_tanh and rnn_relu are supported')
+parser.add_argument('--dtype', type=str, default='float32',
+                    help='if float16 is provided AMP convert model'
+                         'is used to convert model to mixed precision model'
+                         'before running inference')
```
Comment (samskalicky) [inline example/rnn/bucketing/cudnn_rnn_bucketing.py]: what are the supported values for dtype? only FP16?

Comment (samskalicky) [reply example/rnn/bucketing/cudnn_rnn_bucketing.py]: maybe we should define the possible values in the hlep message [float16, float32] so that its clear we dont support int8 yet?

Comment (anirudh2290) [reply example/rnn/bucketing/cudnn_rnn_bucketing.py]: sure! i have updated the help

File: tests/python/gpu/test_contrib_amp.py [hunk 5]
```diff
@@ -300,10 +303,42 @@ def check_amp_convert_hybrid_block():
...
+        result_model = amp.convert_bucketing_module(model)
+        val_sent = []
+        batch_size = 128
+        invalid_label = -1
+        num_sentence = 1000
+        buckets = [5, 10, 20, 30, 40]
+        len_vocab = 50
+
+        for _ in range(num_sentence):
+            len_sentence = randint(6, max(buckets)-1) # leave out the two last buckets empty
```
Comment (ChaiBapchya) [inline tests/python/gpu/test_contrib_amp.py]: Why are the last two buckets left empty?

File: tests/python/train/test_bucketing.py [hunk 6]
```diff
@@ -101,9 +108,14 @@ def sym_gen(seq_len):
         num_epoch=num_epochs,
         batch_end_callback=mx.callback.Speedometer(batch_size, 50))
     logging.info('Finished fit...')
+    return model
+
+
+def test_bucket_module():
     # This test forecasts random sequence of words to check bucketing.
     # We cannot guarantee the accuracy of such an impossible task, and comments out the following line.
```
Comment (ChaiBapchya) [inline tests/python/train/test_bucketing.py]: not related to your PR. But what is the purpose of this line? Can it be made clearer?
Is it just a Note? then "comments out the following line" is not even needed!

Comment (anirudh2290) [reply tests/python/train/test_bucketing.py]: I think this comment is for 117 line which is commented.

File: python/mxnet/module/bucketing_module.py [hunk 7]
```diff
@@ -170,7 +174,7 @@ def get_params(self):
         `(arg_params, aux_params)`
             A pair of dictionaries each mapping parameter names to NDArray values.
         """
-        assert self.binded and self.params_initialized
+        assert self.params_initialized
```
[the remark this answers is not in the document]
Comment (anirudh2290) [reply python/mxnet/module/bucketing_module.py]: This is because get_params() can be called even before bind, for example Module.load followed by get_params.

File: tests/python/gpu/test_contrib_amp.py (hunk 5 above)
Comment (anirudh2290) [reply tests/python/gpu/test_contrib_amp.py]: i think its just to reduce number of binds and make the test faster, I used the same data prep step as in the example.
